// redux/admin-redux/actions.js
import Http from 'utils/Http'

const types = {
  SET_MESSAGE: 'SET_MESSAGE',
  SET_GALLERIES: 'SET_GALLERIES',
  SET_CURRENT_GALLERY: 'SET_CURRENT_GALLERY',
  SET_GALLERIES_LISTS: 'SET_GALLERIES_LISTS',
  SET_CURRENT_GALLERIES_LIST: 'SET_CURRENT_GALLERIES_LIST',
  SET_PROGRESS_UPLOAD: 'SET_PROGRESS_UPLOAD',
  SET_UPLOAD_SIZE: 'SET_UPLOAD_SIZE'
}

const setMessage = message => {
  return {
    type: types.SET_MESSAGE,
    message
  }
}

const setGalleries = galleries => {
  return {
    type: types.SET_GALLERIES,
    galleries
  }
}

const setCurrentGallery = gallery => {
  return {
    type: types.SET_CURRENT_GALLERY,
    gallery
  }
}

const setGalleriesLists = galleriesLists => {
  return {
    type: types.SET_GALLERIES_LISTS,
    galleriesLists
  }
}

const setCurrentGalleriesList = galleriesList => {
  return {
    type: types.SET_CURRENT_GALLERIES_LIST,
    galleriesList
  }
}

const setProgressUpload = progressValue => {
  return {
    type: types.SET_PROGRESS_UPLOAD,
    progressValue
  }
}

const setUploadSize = uploadSizeValue => {
  return {
    type: types.SET_UPLOAD_SIZE,
    uploadSizeValue
  }
}

const getGalleries = () => {
  return function (dispatch) {
    return Http.get('/galleries')
      .then(data => {
        dispatch(setGalleries(data.galleries))
      })
  }
}

const deletePicture = pictureId => {
  return function (dispatch) {
    return Http.delete(`/picture/${pictureId}`)
      .then(data => dispatch(setCurrentGallery(data.gallery)))
  }
}

const setPictureGalleryPreview = pictureId => {
  return function (dispatch) {
    return Http.put(`/picture/${pictureId}`, { galleryPreview: true })
      .then(data => dispatch(setCurrentGallery(data.gallery)))
  }
}

const addPictures = (galleryId, files) => {
  return function (dispatch) {
    const onProgress = event => {
      dispatch(setProgressUpload(event.loaded))
      dispatch(setUploadSize(event.total))
    }

    const formData = new FormData()
    formData.append("galleryId", galleryId)
    Array.from(files).forEach((file, index) => formData.append(index, file))

    return Http.post(`/picture`, formData, null, onProgress)
      .then(data => {
        dispatch(setProgressUpload(0))
        dispatch(setUploadSize(0))
        dispatch(setCurrentGallery(data.gallery))
        return data
      })
  }
}

const putGallery = (galleryId, body) =>
  function (dispatch) {
    return Http.put(`/gallery/${galleryId}`, body)
      .then(data => {
        return data
      })
      .then(data => dispatch(setCurrentGallery(data.gallery)))
  }

const postGallery = (body) => {
  return function (dispatch) {
    const onProgress = event => {
      dispatch(setProgressUpload(event.loaded))
      dispatch(setUploadSize(event.total))
    }
    const formData = new FormData()
    formData.append('name', body.name)
    formData.append('description', body.description)
    Array.from(body.files).forEach((file, index) => formData.append(index, file))
    return Http.post(`/gallery`, formData, null, onProgress)
      .then(data => {
        dispatch(setProgressUpload(0))
        dispatch(setUploadSize(0))
        dispatch(setGalleries(data.galleries))
        return data
      })
  }
}

const deleteGallery = (galleryId) =>
  function (dispatch) {
    return Http.delete(`/gallery/${galleryId}`)
      .then(data => dispatch(setGalleries(data.galleries)))
  }

const getGalleriesLists = () =>
  function (dispatch) {
    Http.get('/gallerieslists')
      .then(galleriesLists => dispatch(setGalleriesLists(galleriesLists)))
  }

const postGalleriesList = (body) => {
  return function (dispatch) {
    const onProgress = event => {
      dispatch(setProgressUpload(event.loaded))
      dispatch(setUploadSize(event.total))
    }
    const formData = new FormData()
    formData.append('name', body.name)
    formData.append('galleries', JSON.stringify(body.galleriesForms.map(galleryForm => galleryForm.name)))
    body.galleriesForms.forEach(galleryForm => {
      Array.from(galleryForm.files).forEach((galleryFiles, index) => formData.append(`${galleryForm.name}${index}`, galleryFiles))
    })
    return Http.post('/gallerieslist', formData, null, onProgress)
      .then(galleriesLists => {
        dispatch(setProgressUpload(0))
        dispatch(setUploadSize(0))
        dispatch(setGalleriesLists(galleriesLists))
        return galleriesLists
      })
  }
}

const putGalleriesList = (id, body) => {
  return function (dispatch) {
    return Http.put(`/gallerieslist/${id}`, body)
      .then(galleriesLists => dispatch(setGalleriesLists(galleriesLists)))
  }
}

const deleteGalleriesList = id => {
  return function (dispatch) {
    return Http.delete(`/gallerieslist/${id}`)
      .then(data => {
        dispatch(setCurrentGalleriesList(null))
        dispatch(setGalleriesLists(data.galleriesLists))
      })
  }
}

const addGalleryToList = (id, body) => {
  return function (dispatch) {
    return Http.put(`/gallerieslist/addGallery/${id}`, body)
      .then(galleriesList => dispatch(setCurrentGalleriesList(galleriesList)))
  }
}

const removeGalleryToList = (id, body) => {
  return function (dispatch) {
    return Http.put(`/gallerieslist/removeGallery/${id}`, body)
      .then(galleriesList => dispatch(setCurrentGalleriesList(galleriesList)))
  }
}

const updateDB = () => {
  return function (dispatch) {
    Http.get('/updatedb')
      .then(
        response => response.json(),
        error => console.log('An error occured.', error)
      )
      .then(
        json => {
          dispatch(setMessage(json.update))
        }
      )
  }
}

export {
  types,
  updateDB,
  getGalleries,
  setCurrentGallery,
  deletePicture,
  setPictureGalleryPreview,
  addPictures,
  setProgressUpload,
  setUploadSize,
  putGallery,
  postGallery,
  deleteGallery,
  addGalleryToList,
  removeGalleryToList,
  getGalleriesLists,
  postGalleriesList,
  putGalleriesList,
  deleteGalleriesList,
  setCurrentGalleriesList
}
