// redux/gallery-redux/actions.js
import Http from 'utils/Http'

import { number_of_pictures } from './constants'

const types = {
  ADD_GALLERY_PICTURES: 'ADD_GALLERY_PICTURES',
  SET_GALLERY: 'SET_GALLERY',
  SET_IS_FETCHING: 'SET_IS_FETCHING',
  SET_CURRENT_PICTURE_INDEX: 'SET_CURRENT_PICTURE_INDEX',
  LOAD_MORE_PICTURES: 'LOAD_MORE_PICTURES'
}

function addGalleryPictures(pictures) {
  return {
    type: types.ADD_GALLERY_PICTURES,
    pictures
  }
}

function setGallery(gallery) {
  return {
    type: types.SET_GALLERY,
    gallery
  }
}

function setIsFetching(bool) {
  return {
    type: types.SET_IS_FETCHING,
    bool
  }
}

function setCurrentPictureIndex(id) {
  return {
    type: types.SET_CURRENT_PICTURE_INDEX,
    id
  }
}

function loadMorePictures() {
  return {
    type: types.LOAD_MORE_PICTURES
  }
}

const getPictures = galleryId =>
  galleryId != null
    ? fetchGallery(galleryId)
    : fetchPictures()

const fetchGallery = galleryId => {
  return function (dispatch) {
    dispatch(setIsFetching(true))
    return Http.get(`/gallery/${galleryId}`)
      .then(json => {
        return json
      })
      .then(gallery => {
        dispatch(setGallery({ ...gallery, pictures: [] }))
        // loadPictures(gallery, dispatch)
        dispatch(addGalleryPictures(gallery.pictures.slice(0, number_of_pictures)))
        return gallery
      })
      .then(gallery => {
        dispatch(setGallery(gallery))
        dispatch(setIsFetching(false))
      })
  }
}

const loadPictures = async (gallery, dispatch) => {
  const picturesLoaded = []
  let nbOfPictures = 0
  gallery.pictures.forEach(
    (picture, index) => {
      Http.get(picture.addr, null, 'blob')
        .then(pictureLoaded => {
          picture.addr = URL.createObjectURL(pictureLoaded)
          picturesLoaded.push(picture)
          nbOfPictures += 1
          if (picturesLoaded.length % 20 === 0 || nbOfPictures === gallery.pictures.length) {
            dispatch(addGalleryPictures(picturesLoaded))
            if (nbOfPictures === gallery.pictures.length) {
              dispatch(setIsFetching(false))
            }
            picturesLoaded.splice(0, picturesLoaded.length - 1)
          }
          return picture
        })
    }
  )
  return picturesLoaded
}

const fetchPictures = () => {
  return function (dispatch) {
    dispatch(setIsFetching(true))
    return Http.get(`/pictures`)
      .then(json => {
        json.name = 'Toutes les photos'
        dispatch(setGallery(json))
        // loadPictures(json, dispatch)
        dispatch(addGalleryPictures(json.pictures.slice(0, number_of_pictures)))
        dispatch(setIsFetching(false))
      })
  }
}

export {
  types,
  addGalleryPictures,
  setGallery,
  loadMorePictures,
  setCurrentPictureIndex,
  getPictures
}
