// nav-redux/actions.js
import config from 'config'

const types = {
  SET_GALLERIES: 'SET_GALLERIES',
  SET_GALLERIES_LISTS: 'SET_GALLERIES_LISTS',
  SET_CURRENT_GALLERIES: 'SET_CURRENT_GALLERIES',
  SET_GALLERIES_NOT_IN_LISTS: 'SET_GALLERIES_NOT_IN_LISTS'
}

function setGalleries (galleries) {
  return {
    type: types.SET_GALLERIES,
    galleries
  }
}

function setGalleriesLists (lists) {
  return {
    type: types.SET_GALLERIES_LISTS,
    lists
  }
}

function setCurrentGalleries(galleries) {
  return {
    type: types.SET_CURRENT_GALLERIES,
    galleries
  }
}

function setGalleriesNotInLists (galleries) {
  return {
    type: types.SET_GALLERIES_NOT_IN_LISTS,
    galleries
  }
}

function fetchGalleries () {
  return dispatch => {
    fetch(config.adressServer+'/galleries')
    .then(
      response => response.json(),
      error => console.log('An error occured.', error)
    )
    .then(
      ({galleries}) => dispatch(setGalleries(galleries))
    )
  }
}

function fetchGalleriesLists () {
  return dispatch => {
    fetch(config.adressServer+'/gallerieslists')
    .then(
      response => response.json(),
      error => console.log('An error occured.', error)
    )
    .then(

      galleriesLists => {
        dispatch(setGalleriesLists(galleriesLists))}
    )
  }
}

function fetchGalleriesNotInLists () {
  return dispatch => {
    fetch(config.adressServer+'/galleriesnotinlists')
    .then(
      response => response.json(),
      error => console.log('An error occured.', error)
    )
    .then(
      galleries => {
        dispatch(setGalleriesNotInLists(galleries))}
    )
  }
}

export {
  types,
  fetchGalleries,
  fetchGalleriesLists,
  fetchGalleriesNotInLists,
  setCurrentGalleries
}
